Support RayJob InTreeAutoscaling: update RayCluster.IsTopLevel() and RayJob.Skip() (0.14 branch)

hiboyang opened the pull request:
#### What type of PR is this?

/kind feature

**This PR is replaced** by https://github.com/kubernetes-sigs/kueue/pull/8082, which is created based on `main` branch. We will continue the discussion in that new PR (https://github.com/kubernetes-sigs/kueue/pull/8082),

#### What this PR does / why we need it:

This PR is to provide code change to help discussion in
[Support RayJob InTreeAutoscaling](https://github.com/kubernetes-sigs/kueue/issues/7605)

#### Which issue(s) this PR fixes:

[Support RayJob InTreeAutoscaling](https://github.com/kubernetes-sigs/kueue/issues/7605)

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

Yes, this PR tries to support RayJob InTreeAutoscaling, which was not supported previously.

mimowo commented:
/ok-to-test

yaroslava-serdiuk commented on pkg/controller/jobs/rayjob/rayjob_controller.go:
```diff
@@ -104,7 +105,15 @@ func (j *RayJob) Suspend() {
+		return true
+	}
+	// Short term solution to support RayJob InTreeAutoscaling: https://github.com/kubernetes-sigs/kueue/issues/7605
+	if ptr.Deref(j.Spec.RayClusterSpec.EnableInTreeAutoscaling, false) &&
```
```suggestion
	if ptr.Deref(j.Spec.RayClusterSpec.EnableInTreeAutoscaling, true) &&
```
IIUC we should skip if the autoscaling is enabled

Singularity23x0 commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -630,7 +630,16 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
+			log.V(10).Info("Not using PodLabelSelector from job")
+		}
+
+		return ctrl.Result{}, workloadslicing.StartWorkloadSlicePods(ctx, r.client, object, podLabelSelector)
```
Is it possible that the label selector we get here is empty by design? Meaning we would want to StartWorkloadSlicePods with any set of labels? 
If yes then the current logic may clash with this case due to defaulting to using the owner reference UID if the selector is empty.

hiboyang replied:
Updated PR, do not need logic using pod selector anymore, can still use owner reference UID.

hiboyang commented on pkg/controller/jobs/rayjob/rayjob_controller.go:
```diff
@@ -104,7 +105,15 @@ func (j *RayJob) Suspend() {
+		return true
+	}
+	// Short term solution to support RayJob InTreeAutoscaling: https://github.com/kubernetes-sigs/kueue/issues/7605
+	if ptr.Deref(j.Spec.RayClusterSpec.EnableInTreeAutoscaling, false) &&
```
Working on it

hiboyang commented:
/ok-to-test

hiboyang commented:
/ok-to-test

hiboyang commented:
/ok-to-test

mimowo commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -374,7 +374,28 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
+				log.V(10).Info("Ancestor job not converted to generic job", "ancestorJobName", ancestorJob.GetName())
+				ancestorWorkload, err = r.getWorkloadForObject(ctx, ancestorJob)
+			}
+			if err != nil {
```
Are these code changes needed? Let me know what is the reason for introducing. I would think implementing the IsTopLevel is enough here.

mimowo replied:
This has some issues:
1. the request for the submitter Job is not represented by Kueue Workload, so bypassing quota reservation, this may lead to subtle bugs for TAS
2. it seems incompatible with LocalQueueDefaulting, because then Workload will be created for the submitter Job anyway, consequences are hard to predict
3. it is incompatible with the configuration setting for `manageJobsWithoutQueue: true`, then it will just suspend the submitter Job

I think it will be better to manager the submitter Job as another topLevelJob, and create a Workload for it. 

To achieve that we can implement the IsTopLevel for Job to lookup if parent it RayJob with autoscaling enabled. 

Long term I think integration for RayJob could work this way by default - to create two workloads, one for submitter Job, one for RayCluster.

hiboyang commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -374,7 +374,28 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
+			ancestorGenericJob := CreateGenericJobFromRuntimeObject(ancestorJob)
+			if ancestorGenericJob != nil {
+				log.V(10).Info("Ancestor job converted to generic job, trying to convert it again to JobWithCustomWorkloadRetriever")
+				jobWithCustomWorkloadRetriever, ok := ancestorGenericJob.(JobWithCustomWorkloadRetriever)
```
> From what I read the submitter job is a mechanism to inject pods to the RayCluster. The RayJob is ancestor to the RayCluster, the RayJob usually doesn't have ancestor. Unless there is a specific scenario when user create a specific Job that creates RayJobs.
> 
> So for the change you provided:
> 
> 1. if the ReconcileGenericJob is run for RayCluster: RayCluster has isTopLevel() method that return true and so it matches with workload that kueue created for RayCluster or it will create a workload if the workload wasn't created before.
> 2. if the ReconcileGenericJob is run for RayJob: RayJob doesn't implement isTopLevel(), so then ancestor is searched which is usually nil and so RayJob will be treated as topLevel. And so there willl be another workload created for RayJob.
> 
> If we have different workloads created for RayJob and RayCluster it will lead to quota issues.

@yaroslava-serdiuk are you raising questions or want me to clarify? The `JobWithCustomWorkloadRetriever` is used in following scenario:
1. Submit a RayJob with InTreeAutoscaling:true
2. RayJob is skipped due to Skip() method and does not create workload
3. RayJob creates RayCluster, which creates workload
4. RayJob creates submitter job (batch/Job). This is Ray internal logic.
5. The submitter job (batch/Job) triggers reconcile in Kueue. The submitter's ancestor is RayJob. Now RayJob implements `JobWithCustomWorkloadRetriever` and will get workload correctly. Without it, the submitter job will get workload from its ancestor (RayJob) by `ownerReferences`, which is wrong, since the workload is owned by underlying RayCluster (not RayJob).

yaroslava-serdiuk replied:
I see, thanks for clarification.

hiboyang replied:
If we exclude RayJob in config map and the ancestorJob is nil for submitter job, the submitter job will become top level job due code like `isTopLevelJob = ancestorJob == nil`, will this cause issue?

yaroslava-serdiuk replied:
Actually I see that submitter job will never have a [kueue label](https://github.com/ray-project/kuberay/blob/master/ray-operator/controllers/ray/rayjob_controller.go#L673), so if the ManageJobsWithoutQueueName is false (which should be), than kueue won't manage this job. 

So the changes you added to handle submitter job are not needed.

yaroslava-serdiuk replied:
If we exclude RayJob in config map, than there would be no need to implement isTopLevel() for RayCluster. However, since we do not currently plan for RayJob deprecation, I would leave the implementation as it is.

hiboyang commented:
/ok-to-test

hiboyang commented:
/ok-to-test

hiboyang commented:
/ok-to-test

mbobrovskyi commented:
You are trying to merge to release-0.14 brach. Is it correct?

hiboyang commented:
@mbobrovskyi I got a lot of conflicting when trying to rebase this PR on main branch. Thus created another PR based on main branch: https://github.com/kubernetes-sigs/kueue/pull/8082. Would you help to review/merge the new PR: https://github.com/kubernetes-sigs/kueue/pull/8082?

mimowo commented:
@hiboyang that is ok, but it means losing discussion history. I think it should also be possible to change the target branch for this PR from "release-0.14" to "main" here: 
<img width="1243" height="150" alt="image" src="https://github.com/user-attachments/assets/d3451f38-bf84-4ec6-a0f1-eaf7a3e96ec0" />

Let us know your preference. I'm ok to also move my comments there.

mimowo commented on pkg/controller/jobframework/interface.go:
```diff
@@ -119,6 +119,13 @@ type JobWithCustomValidation interface {
+// JobWithCustomWorkloadRetriever interface should be implemented by jobs which have
+// their own logic to get workload, e.g. for RayJob with InTreeAutoScaling, the workload
+// is attached on the underlying RayCluster, RayJob needs to get workload through that RayCluster
+type JobWithCustomWorkloadRetriever interface {
```
I think this interface is no longer needed if we make the submitter Job top-level as mentioned here: https://github.com/kubernetes-sigs/kueue/pull/7769/files#r2588885609

hiboyang commented:
> @hiboyang that is ok, but it means losing discussion history. I think it should also be possible to change the target branch for this PR from "release-0.14" to "main" here: <img alt="image" width="1243" height="150" src="https://private-user-images.githubusercontent.com/10359181/522848292-d3451f38-bf84-4ec6-a0f1-eaf7a3e96ec0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQ5NTY1ODQsIm5iZiI6MTc2NDk1NjI4NCwicGF0aCI6Ii8xMDM1OTE4MS81MjI4NDgyOTItZDM0NTFmMzgtYmY4NC00ZWM2LWEwZjEtZWFmN2EzZTk2ZWMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjA1VDE3MzgwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNjNWUxMDdkYjQzODNiNjZiNTg1MGY5NTUyNDc3ZGJlMDk2ZGE5ODM2YTE0NjI3YjMxMDY1MWVjNjZkOWZkOGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.dqVfA72iP-9ZdksvlcSgvlIjvJ8Csb0-tGBS9EmA9Kc">
> 
> Let us know your preference. I'm ok to also move my comments there.

I tried changing the target branch to be `main`, then this PR was messed up with a lot of changes from different places. Thus I reverted it back to set the target branch back to be `release-0.14`. We can keep discussion in that new https://github.com/kubernetes-sigs/kueue/pull/8082. Let me copy unaddressed comments to there from this PR and continue the discussion there.
